tokio/src/sync/broadcast.rs

```diff
@@ -372,11 +372,43 @@ struct Recv<'a, T> {
 
     /// Entry in the waiter `LinkedList`.
     waiter: UnsafeCell<Waiter>,
+
+    // Flag that indicates whether to output the number
+    // of sent messages that the `Receiver` behind this `Recv`
+    // has yet to receive.
+    output_kind: OutputQueueStatus,
```

Comment:
This is a very complicated change. Is there a reason this can't be done in a simpler way?

Reply:
I don't think it's that complicated. What appears to make this complicated is that we need to wrap the return type of `recv_ref` and `poll` in enum types to get around the limitation of the generic type in `Receiver` not having a `Clone` bound. Other than that we basically just check this flag in `recv_ref` to determine whether to add the queue status.

Reply:
Sorry, I still think this is way way too complicated. Your diff is 260 lines, but could be replaced with the following four-line change:
```Rust
fn len(&self) -> usize {
    let pos = self.shared.tail.lock().pos;
    pos - self.next
}
```
Even if you want to avoid locking it twice (which I am not convinced is necessary), you could just have `recv_ref` always return the position when it returns a message and ignore it for the existing send methods. There are also solutions that involve moving the `pos` field to an atomic outside the mutex.

Reply:
That's fair. I can try the approach using an atomic on the `pos` field later.